data_lang/pretty.py

```diff
@@ -1,40 +1,546 @@
+        return self._Surrounded("(BashAssoc", " ", self._Join(mdocs, "", " "),
+                                ")")
+
+    def _Value(self, val):
+        # type: (value_t) -> MeasuredDoc
+
+        with tagswitch(val) as case:
+            if case(value_e.Null):
+                return self._Styled(NULL_STYLE, _Text("null"))
+
+            elif case(value_e.Bool):
+                b = cast(value.Bool, val).b
+                return self._Styled(BOOL_STYLE,
+                                    _Text("true" if b else "false"))
+
+            elif case(value_e.Int):
+                i = cast(value.Int, val).i
+                return self._Styled(NUMBER_STYLE, _Text(mops.ToStr(i)))
+
+            elif case(value_e.Float):
+                f = cast(value.Float, val).f
+                return self._Styled(NUMBER_STYLE, _Text(str(f)))
+
+            elif case(value_e.Str):
+                s = cast(value.Str, val).s
+                return self._StringLiteral(s)
+
+            elif case(value_e.Range):
+                r = cast(value.Range, val)
+                return self._Styled(
+                    NUMBER_STYLE,
+                    _Concat([
+                        _Text(str(r.lower)),
+                        _Text(" .. "),
+                        _Text(str(r.upper))
+                    ]))
+
+            elif case(value_e.List):
+                vlist = cast(value.List, val)
+                heap_id = HeapValueId(vlist)
+                if self.visiting.get(heap_id, False):
+                    return _Concat([
+                        _Text("["),
+                        self._Styled(CYCLE_STYLE, _Text("...")),
+                        _Text("]")
+                    ])
+                else:
+                    self.visiting[heap_id] = True
+                    result = self._YshList(vlist)
+                    self.visiting[heap_id] = False
+                    return result
+
+            elif case(value_e.Dict):
+                vdict = cast(value.Dict, val)
+                heap_id = HeapValueId(vdict)
+                if self.visiting.get(heap_id, False):
+                    return _Concat([
+                        _Text("{"),
+                        self._Styled(CYCLE_STYLE, _Text("...")),
+                        _Text("}")
+                    ])
+                else:
+                    self.visiting[heap_id] = True
+                    result = self._YshDict(vdict)
+                    self.visiting[heap_id] = False
+                    return result
+
+            elif case(value_e.BashArray):
+                varray = cast(value.BashArray, val)
+                return self._BashArray(varray)
+
+            elif case(value_e.BashAssoc):
+                vassoc = cast(value.BashAssoc, val)
+                return self._BashAssoc(vassoc)
+
+            else:
+                ysh_type = value_str(val.tag(), dot=False)
+                id_str = ValueIdString(val)
+                return _Text("<" + ysh_type + id_str + ">")
```

Comment:
Are there any other values cases we should handle? E.g. I added a case for `range`.

Reply:
We might want to, but we can add them later